tensorflow_addons/optimizers/gradient_accumulator.py

```diff
@@ -12,228 +12,91 @@
+        self.variable_dtype = policy.variable_dtype
+        self._accu_gradients = {
+            var.ref(): self.optimizer.add_slot(var, "ga") for var in trainable_variables
+        }
+
+    def __call__(self, grads_and_vars, *args, **kwargs):
+
+        variables = [var for (_, var) in grads_and_vars]
+        accu_gradients = self._accu_gradients
+        step_inc_op = self.step.assign_add(1, read_value=False)
+
+        with tf.control_dependencies([step_inc_op]):
+            can_apply = tf.cast(
+                self.step % self.accu_steps == 0, dtype=self.variable_dtype
             )
-        self._step = variable
-        self._weights.append(self._step)
-
-    @property
-    def gradients(self):
-        """The accumulated gradients on the current replica."""
-        if not self._gradients:
-            raise ValueError(
-                "The accumulator should be called first to initialize the gradients"
+            accumulate = tf.cast(
+                self.step % (self.accu_steps + 1) != 0, dtype=self.variable_dtype
             )
-        return list(
-            gradient.read_value() if gradient is not None else gradient
-            for _, gradient in self._gradients
-        )
 
-    def apply_gradients(self, grads_and_vars, name=None, **kwargs):
-        train_op = self._optimizer.apply_gradients(grads_and_vars, name, **kwargs)
-        with tf.control_dependencies([train_op]):
-            with tf.control_dependencies(
-                [
-                    self._optimizer.iterations.assign_add(
-                        tf.cast(self.step % self._accum_steps == 0, tf.int64),
-                        read_value=False,
-                    )
-                ]
-            ):
-                return self.step.assign_add(1, read_value=False)
-
-    def reset(self):
-        """Resets the accumulated gradients on the current replica."""
-        assign_ops = []
-        if not self._gradients:
-            return assign_ops
-
-        for _, gradient in self._gradients:
-            if gradient is not None:
-                assign_ops.append(
-                    gradient.assign(
-                        tf.zeros_like(gradient),
-                        use_locking=self._use_locking,
-                        read_value=False,
-                    )
-                )
-
-        return tf.group(assign_ops)
+        accum_ops = list()
+        for grad, var in grads_and_vars:
 
-    @property
-    def inner_optimizer(self):
-        """The optimizer that this LossScaleOptimizer is wrapping."""
-        return self._optimizer
+            # Get the accumulated gradient
+            grad_accum = accu_gradients[var.ref()] * accumulate
 
-    @property
-    def iterations(self):
-        return self._optimizer.iterations
+            if isinstance(grad, tf.IndexedSlices):
+                added = tf.IndexedSlices(
+                    values=grad.values + tf.gather_nd(grad_accum, grad.indices),
```

Comment:
@fsx950223 This here is interesting:

I have to write `grad.indices[..., None]` for the `Embedding` layer to work - otherwise it crashes.